libbeat/common/transport/tlscommon/tls_config.go

```diff
@@ -86,68 +87,177 @@ func (c *TLSConfig) ToConfig() *tls.Config {
+			if len(cs.PeerCertificates) == 0 {
+				return fmt.Errorf("no peer certificates")
+			}
+
+			dnsnames := cs.PeerCertificates[0].DNSNames
+			var serverName string
+			if len(dnsnames) == 0 || len(dnsnames) == 1 && dnsnames[0] == "" {
+				serverName = cs.PeerCertificates[0].Subject.CommonName
+			} else {
+				serverName = dnsnames[0]
+			}
+			if len(serverName) > 0 && len(cs.ServerName) > 0 && serverName != cs.ServerName {
+				return x509.HostnameError{Certificate: cs.PeerCertificates[0], Host: cs.ServerName}
+			}
+			opts := x509.VerifyOptions{
+				Roots:         cfg.RootCAs,
+				Intermediates: x509.NewCertPool(),
+			}
+			for _, cert := range cs.PeerCertificates[1:] {
+				opts.Intermediates.AddCert(cert)
+			}
+			verifiedChains, err := cs.PeerCertificates[0].Verify(opts)
+			if err != nil {
+				return err
+			}
+
+			if pin {
+				return verifyCAPin(cfg.CASha256, verifiedChains)
+			}
+			return nil
 		}
-	}
+	case VerifyCertificate:
+		return func(cs tls.ConnectionState) error {
+			// On the client side, PeerCertificates can't be empty.
+			if len(cs.PeerCertificates) == 0 {
+				return fmt.Errorf("no peer certificates")
+			}
 
-	if pin && skipHostName {
-		return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-			_, _, err := verifyCertificateExceptServerName(rawCerts, cfg)
+			opts := x509.VerifyOptions{
+				Roots:         cfg.RootCAs,
+				Intermediates: x509.NewCertPool(),
+			}
+			for _, cert := range cs.PeerCertificates[1:] {
+				opts.Intermediates.AddCert(cert)
+			}
+			verifiedChains, err := cs.PeerCertificates[0].Verify(opts)
 			if err != nil {
 				return err
 			}
-			return verifyCAPin(cfg.CASha256, verifiedChains)
+
+			if pin {
+				return verifyCAPin(cfg.CASha256, verifiedChains)
+			}
+			return nil
 		}
+	case VerifyStrict:
+		if pin {
+			return func(cs tls.ConnectionState) error {
+				return verifyCAPin(cfg.CASha256, cs.VerifiedChains)
+			}
+		}
+	default:
 	}
 
-	if !pin && skipHostName {
-		return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-			_, _, err := verifyCertificateExceptServerName(rawCerts, cfg)
+	return nil
+
+}
+
+func makeVerifyServerConnection(cfg *TLSConfig) func(tls.ConnectionState) error {
+	switch cfg.Verification {
+	case VerifyFull:
+		return func(cs tls.ConnectionState) error {
+			if len(cs.PeerCertificates) == 0 {
+				if cfg.ClientAuth == tls.RequireAndVerifyClientCert {
+					return fmt.Errorf("no peer certificates")
+				}
+				return nil
+			}
+
+			dnsnames := cs.PeerCertificates[0].DNSNames
+			var serverName string
+			if len(dnsnames) == 0 || len(dnsnames) == 1 && dnsnames[0] == "" {
+				serverName = cs.PeerCertificates[0].Subject.CommonName
+			} else {
+				serverName = dnsnames[0]
+			}
+			if len(serverName) > 0 && len(cs.ServerName) > 0 && serverName != cs.ServerName {
+				return x509.HostnameError{Certificate: cs.PeerCertificates[0], Host: cs.ServerName}
+			}
```

Comment:
peer "host" verification is duplicate from client code. How about moving it into a separate testable function?

Do we test with IPs as well?